server/filestore.go

```diff
@@ -3791,6 +3869,166 @@ func (fs *fileStore) writeMsgRecord(seq uint64, ts int64, subj string, hdr, msg
+		// returned and the algorithm defaults to no compression.
+		return fmt.Errorf("failed to read existing metadata header: %w", err)
+	}
+	if meta.Algorithm == alg {
+		// The block is already compressed with the chosen algorithm so there
+		// is nothing else to do.
+		return nil
+	} else if alg != NoCompression {
+		// The block is already compressed using some algorithm, so we need
+		// to decompress the block using the existing algorithm before we can
+		// recompress it with the new one.
+		if origBuf, err = meta.Algorithm.Decompress(origBuf); err != nil {
+			return fmt.Errorf("failed to decompress original block: %w", err)
+		}
+	}
+
+	// Rather than modifying the existing block on disk (which is a dangerous
+	// operation if something goes wrong), create a new temporary file. We will
+	// write out the new block here and then swap the files around afterwards
+	// once everything else has succeeded correctly.
+	tmpFD, err := os.OpenFile(tmpFN, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, defaultFilePerms)
+	if err != nil {
+		return fmt.Errorf("failed to create temporary file: %w", err)
+	}
+
+	// The original buffer at this point is uncompressed, so we will now compress
+	// it if needed. Note that if the selected algorithm is NoCompression, the
+	// Compress function will just return the input buffer unmodified.
+	cmpBuf, err := alg.Compress(origBuf)
+	if err != nil {
+		return fmt.Errorf("failed to compress block: %w", err)
+	}
+
+	// We only need to write out the metadata header if compression is enabled.
+	// If we're trying to uncompress the file on disk at this point, don't bother
+	// writing metadata.
+	if alg != NoCompression {
+		meta := &CompressionInfo{
+			Algorithm:    alg,
+			OriginalSize: uint64(len(origBuf)),
+		}
+		cmpBuf = append(meta.MarshalMetadata(), cmpBuf...)
+	}
+
+	// Re-encrypt the block if necessary.
+	if mb.bek != nil && len(cmpBuf) > 0 {
+		bek, err := genBlockEncryptionKey(mb.fs.fcfg.Cipher, mb.seed, mb.nonce)
+		if err != nil {
+			return err
+		}
+		mb.bek = bek
+		mb.bek.XORKeyStream(cmpBuf, cmpBuf)
+	}
+
+	// Write the new block data (which might be compressed or encrypted) to the
+	// temporary file.
+	if n, err := tmpFD.Write(cmpBuf); err != nil {
+		return fmt.Errorf("failed to write to temporary file: %w", err)
+	} else if n != len(cmpBuf) {
+		return fmt.Errorf("short write to temporary file (%d != %d)", n, len(cmpBuf))
+	}
+	if err := tmpFD.Sync(); err != nil {
+		return fmt.Errorf("failed to sync temporary file: %w", err)
+	}
+	if err := tmpFD.Close(); err != nil {
+		return fmt.Errorf("failed to close temporary file: %w", err)
+	}
+
+	// Now swap the files around. First we'll move the original file aside,
+	// then we'll move the temporary file into place, and then finally we will
+	// delete the original file.
+	if err := os.Rename(origFN, oldFN); err != nil {
+		return fmt.Errorf("failed to move original file aside: %w", err)
+	}
+	if err := os.Rename(tmpFN, origFN); err != nil {
```

Comment:
I think we only want this one? Do we need the one above to move old then remove it?